src/python/pants/option/subsystem.py

```diff
@@ -101,10 +101,11 @@ def register_options(cls, register):
 
         Subclasses may override and call register(*args, **kwargs).
         """
-        for attrname in dir(cls):
-            attr = getattr(cls, attrname)
-            if isinstance(attr, Option):
-                register(*attr.args, **attr.kwargs)
+        # NB: We register these in class attribute order, starting from the base class down.
```

Comment:
Could you please add to this comment explaining that the order does matter, as it impacts how `./pants help` renders for that subsystem?

Reply:
I actually think this is a bit.. I don't find the word for it, but the order seems arbitrary at cases and impedes development at others (stopping to consider in which order to place the options etc) rather than just presenting them in alphabetical order, which would make it easier to look up if you know the name you're looking for. If you don't know the name, you're going to have to scan the whole thing any way, so no real win trying to be clever anticipating which options are likely the most frequent ones being looked up to place at the top.

Reply:
FWIW both topics are somewhat of an implementation detail no? The whole point is to say that declaring the attribute has the same semantics as calling `register` w.r.t. ordering.

If calling `register` in a certain order matters, that's on `register` :smile:

Reply:
(That being said, if we moved this implementation detail out of `register` we can and should and need to totally document it here)

Reply:
See https://github.com/pantsbuild/pants/pull/10466 for where we made the change to not be alphabetical. That principle Benjy describes is particularly relevant to global_options.py.